### PR TITLE
Moved logger back to cerr

### PR DESCRIPTION
**Pull request by LarryOsterman**
# Pull Request Checklist

Please leverage this checklist as a reminder to address commonly occurring feedback when submitting a pull request to make sure your PR can be reviewed quickly:

See the detailed list in the [contributing guide](https://github.com/Azure/azure-sdk-for-cpp/blob/main/CONTRIBUTING.md#pull-requests).

- [ ] [C++ Guidelines](https://azure.github.io/azure-sdk/cpp_introduction.html)
- [ ] Doxygen docs
- [ ] Unit tests
- [ ] No unwanted commits/changes
- [ ] Descriptive title/description
  - [ ] PR is single purpose
  - [ ] Related issue listed
- [ ] Comments in source
- [ ] No typos
- [ ] Update changelog
- [ ] Not work-in-progress
- [ ] External references or docs updated
- [ ] Self review of PR done
- [ ] Any breaking changes?

### CODE REVIEW

**Review by antkmsft — APPROVED** [review]
Thank you!

BTW, I just found another source confirming that stderr is intended for **diagnostic** messages - POSIX standard: https://pubs.opengroup.org/onlinepubs/9699919799/functions/stdin.html ("_...and standard error (for writing diagnostic output)_"), and all logging is diagnostics.

Sorry for misunderstanding with newlines, I missed the logic where we handle newlines (got confused by the unit test expecting strings line `"Informational\n"` when we log `"Informational"`). LGTM!
Still, it won't be possible to include newlines in the log message itself but that is not a very common scenario, and for that `Log::Write()` will be used.

**Comment by antkmsft** [comment]
Can you please also move the static variables to an unnamed namespace? There are no disadvantages in moving them there.

**Comment by antkmsft** [comment]
I also propose to trim `\n` in `sync()` - so that console logger does not need to have that detection logic, and all log listeners, including custom ones, would get all messages without `\n` at the end. Right now, they would get those that were logged via stream with `\n` at the end, while those logged via `Log::Write()` won't have `\n`s.

File: sdk/core/azure-core/src/environment_log_level_listener.cpp [hunk 1]
```diff
@@ -120,15 +120,16 @@ EnvironmentLogLevelListener::GetLogListener()
 
   static std::function<void(Logger::Level level, std::string const& message)> const consoleLogger =
       [](auto level, auto message) {
-        std::ostream* os = &std::cout;
-        if (level == Logger::Level::Error)
+        if (message.empty())
```
Comment (antkmsft) [inline sdk/core/azure-core/src/environment_log_level_listener.cpp:123]: I think `Log::Write()` should be doing this, so we can say if log message is empty, no log level listener is ever going to be invoked.

File: sdk/core/azure-core/src/environment_log_level_listener.cpp [hunk 2]
```diff
@@ -138,11 +139,11 @@ EnvironmentLogLevelListener::GetLogListener()
         // insert unnecessary new lines.
         if (!message.empty() && message.back() == '\n')
         {
-          *os << std::flush;
+          std::cerr << std::flush;
         }
```
Comment (antkmsft) [inline sdk/core/azure-core/src/environment_log_level_listener.cpp:143]: I think this is no longer needed? Message will no longer ever be empty here, and it normally wouldn't end with `\n`.

File: sdk/core/azure-core/src/logger.cpp [hunk 3]
```diff
@@ -54,21 +54,26 @@ void Logger::SetLevel(Logger::Level level) { Log::SetLogLevel(level); }
 
 int Log::LoggerStringBuffer::sync()
 {
-  Log::Write(m_level, str());
+  std::string val{str()};
+  if (!val.empty())
```
Comment (antkmsft) [inline sdk/core/azure-core/src/logger.cpp:58]: I think it is also better to move this check to `Log::Write()`

File: sdk/core/azure-core/test/ut/logging_test.cpp [hunk 4]
```diff
@@ -242,54 +238,54 @@ TEST(Logger, LoggerStream)
...
-    Log::Stream(Logger::Level::Warning) << "Warning" << std::endl;
-    EXPECT_EQ(message, "Warning\n");
+    Log::Stream(Logger::Level::Warning) << "Warning";
+    EXPECT_EQ(message, "Warning");
     message.clear();
 
-    Log::Stream(Logger::Level::Error) << "Error" << std::endl;
-    EXPECT_EQ(message, "Error\n");
+    Log::Stream(Logger::Level::Error) << "Error";
+    EXPECT_EQ(message, "Error");
```
Comment (antkmsft) [inline sdk/core/azure-core/test/ut/logging_test.cpp:269]: ❤️

File: sdk/core/azure-core/src/environment_log_level_listener.cpp (hunk 1 above)
Comment (LarryOsterman) [reply sdk/core/azure-core/src/environment_log_level_listener.cpp:123]: That's a really good question. And yeah, it makes sense to move this to Log::Write instead of the actual low level logger - at the same point it filters on log level.

File: sdk/core/azure-core/src/logger.cpp (hunk 3 above)
Comment (LarryOsterman) [reply sdk/core/azure-core/src/logger.cpp:58]: This case exists to handle the very special case of inserting `std::endl` which calls flush, and then the `Stream` destructor calls flush again (on an empty string). I'll add a comment explaining why it's here.

File: sdk/core/azure-core/README.md [hunk 5]
```diff
@@ -117,7 +117,8 @@ int main()
 
 Note, the listener callback is executed on the same thread as the operation that triggered the log message. 
  It is recommended implementation due the minimal amount of log message processing on the callback thread.
-Where message processing is required, consider implementing in a way that the callback pushes the message string into a thread-safe queue, so that another thread would pick the messages from that queue and handle them.
+Where message processing is required, consider implementing in a way that the callback pushes the message 
+string into a thread-safe queue, so that another thread would pick the messages from that queue and handle them.
```
Comment (ahsonkhan) [inline sdk/core/azure-core/README.md:121]: This is only a line break change and is purely cosmetic. Unlike source code, such changes are unnecessary in markdown, and in fact make it harder to keep up consistently in different IDEs. We should avoid such line breaks and unrelated changes in PRs.

File: sdk/core/azure-core/CHANGELOG.md [hunk 6]
```diff
@@ -10,7 +10,7 @@
 
 ### Other Changes
 
-- The default logger now logs all non-error outputs to `stdout` instead of `stderr`. Log level `Error` is still logged to `stderr`.
+Empty diagnostic messages will no longer be generated.
```
Comment (ahsonkhan) [inline sdk/core/azure-core/CHANGELOG.md:13]: nit: add back the hyphen to create a list of changes
```suggestion
- Empty diagnostic messages will no longer be generated.
```

File: sdk/core/azure-core/inc/azure/core/internal/diagnostics/log.hpp [hunk 7]
```diff
@@ -50,6 +50,19 @@ namespace Azure { namespace Core { namespace Diagnostics { namespace _internal {
...
+      ~Stream() { m_stream.flush(); }
+      Stream(Stream const&) = delete;
+      Stream& operator=(Stream const&) = delete;
+
+      template <typename T> std::ostream& operator<<(T val) { return m_stream << val; }
+
+    private:
+      LoggerStream& m_stream;
+    };
+
```
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/internal/diagnostics/log.hpp:65]: What's the purpose of this class? Please add comments. It seems to be very closely tied to `LoggerStream` with a custom dtor for flushing and the `<<` operator. Could we just merge them both and keep one class, if the separation isn't critical? It doesn't seem to buy us much, unless I am missing some rationale for the separation.

Also, are we ok with a class name as generic as `Stream`.

Comment (LarryOsterman) [reply sdk/core/azure-core/inc/azure/core/internal/diagnostics/log.hpp:65]: This class is how you use the Log class as an iostream target. You say `Log::Stream(<level>) << "Test";` 

There doesn't seem to be any place to add documentation on how to use the Log class, otherwise I would have added it there.
